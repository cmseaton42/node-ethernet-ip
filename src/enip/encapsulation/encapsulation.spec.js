const encapsulation = require("./index");

describe("Encapsulation", () => {
    describe("Command Validator", () => {
        const { validateCommand } = encapsulation;
        const {
            RegisterSession,
            UnregisterSession,
            SendRRData,
            SendUnitData
        } = encapsulation.commands;

        it("Rejects Invalid Commands", () => {
            expect(validateCommand(0x99)).toBeFalsy();
            expect(validateCommand("hello")).toBeFalsy();
            expect(validateCommand(0x02)).toBeFalsy();
        });

        it("Accepts Proper Commands", () => {
            expect(validateCommand(0x66)).toBeTruthy();
            expect(validateCommand(102)).toBeTruthy();
            expect(validateCommand(RegisterSession)).toBeTruthy();
            expect(validateCommand(UnregisterSession)).toBeTruthy();
            expect(validateCommand(SendRRData)).toBeTruthy();
            expect(validateCommand(SendUnitData)).toBeTruthy();
        });
    });

    describe("Status Parser", () => {
        const { parseStatus } = encapsulation;

        it("Rejects Non-Number Inputs", () => {
            expect(() => parseStatus("test")).toThrow();
            expect(() => parseStatus(null)).toThrow();
            expect(() => parseStatus(undefined)).toThrow();
        });

        it("Returns Proper Human Readable String", () => {
            expect(parseStatus(0)).toEqual("SUCCESS");
            expect(parseStatus(0x01)).toEqual(expect.stringContaining("FAIL"));
            expect(parseStatus(1)).toEqual(expect.stringContaining("FAIL"));
            expect(parseStatus(0x45)).toEqual(expect.stringContaining("FAIL"));
        });
    });

    describe("Header Building Utility", () => {
        const {
            header: { build },
            commands: { RegisterSession }
        } = encapsulation;

        it("Builds Correct Encapsulation Buffer", () => {
            const snap = build(RegisterSession, 0x00, [0x01, 0x00, 0x00, 0x00]);
            expect(snap).toMatchSnapshot();
        });
    });

    describe("Header Parsing Utility", () => {
        const {
            header: { parse, build },
            commands: { SendRRData }
        } = encapsulation;

        it("Builds Correct Encapsulation Buffer", () => {
            const data = build(SendRRData, 98705, [0x01, 0x00, 0x00, 0x00]);
            const snap = parse(data);

            expect(snap).toMatchSnapshot();
        });
    });

    describe("Test Encapsulation Generator Functions", () => {
        const { registerSession, unregisterSession, sendRRData, sendUnitData, listIdentity, listServices } = encapsulation;

        it("Register Session Returns Correct Encapsulation String", () => {
            const data = registerSession();

            expect(data).toMatchSnapshot();
        });

        it("Unregister Session Returns Correct Encapsulation String", () => {
            const data = unregisterSession(98705);

            expect(data).toMatchSnapshot();
        });

        it("SendRRData Returns Correct Encapsulation String", () => {
            const data = sendRRData(98705, Buffer.from("hello world"));

            expect(data).toMatchSnapshot();
        });

        it("SendUnitData Returns Correct Encapsulation String", () => {
            const data = sendUnitData(98705, Buffer.from("hello world"), 32145, 456);

            expect(data).toMatchSnapshot();
        });

        it("ListIdentity Returns Correct Encapsulation String", () => {
            const data = listIdentity();

            expect(data).toMatchSnapshot();
        });

        it("ListServices Returns Correct Encapsulation String", () => {
            const data = listServices();

            expect(data).toMatchSnapshot();
        });
    });

    describe("Test Common Packet Format Helper Functions", () => {
        const {
            CPF: { parse, build, isCmd, ItemIDs }
        } = encapsulation;

        it("Invalid CPF Commands causes an Error to be Thrown", () => {
            const { Null, ListIdentity, ConnectionBased, UCMM } = ItemIDs;

            expect(isCmd(Null)).toBeTruthy();
            expect(isCmd(ListIdentity)).toBeTruthy();
            expect(isCmd(ConnectionBased)).toBeTruthy();
            expect(isCmd(UCMM)).toBeTruthy();
            expect(isCmd(0x8001)).toBeTruthy();
            expect(isCmd(0x01)).toBeFalsy();
            expect(isCmd(0x8003)).toBeFalsy();
            expect(isCmd(0xc1)).toBeFalsy();
        });

        it("Build Helper Function Generates Correct Output", () => {
            const test1 = [
                { TypeID: ItemIDs.Null, data: [] },
                { TypeID: ItemIDs.UCMM, data: "hello world" }
            ];

            const test2 = [
                { TypeID: ItemIDs.Null, data: [] },
                { TypeID: ItemIDs.UCMM, data: "hello world" },
                { TypeID: ItemIDs.ConnectionBased, data: "This is a test" }
            ];

            expect(build(test1)).toMatchSnapshot();
            expect(build(test2)).toMatchSnapshot();
        });

        it("Parse Helper Function Generates Correct Output", () => {
            const test1 = build([
                { TypeID: ItemIDs.Null, data: [] },
                { TypeID: ItemIDs.UCMM, data: "hello world" }
            ]);

            const test2 = build([
                { TypeID: ItemIDs.Null, data: [] },
                { TypeID: ItemIDs.UCMM, data: "hello world" },
                { TypeID: ItemIDs.ConnectionBased, data: "This is a test" }
            ]);

            expect(parse(test1)).toMatchSnapshot();
            expect(parse(test2)).toMatchSnapshot();
        });
    });
});
