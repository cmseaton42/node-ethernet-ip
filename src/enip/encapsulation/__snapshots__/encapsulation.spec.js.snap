// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Encapsulation Header Building Utility Builds Correct Encapsulation Buffer 1`] = `
Object {
  "data": Array [
    101,
    0,
    4,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`Encapsulation Header Parsing Utility Builds Correct Encapsulation Buffer 1`] = `
Object {
  "command": "SendRRData",
  "commandCode": 111,
  "data": Object {
    "data": Array [
      1,
      0,
      0,
      0,
    ],
    "type": "Buffer",
  },
  "length": 4,
  "options": 0,
  "session": 98705,
  "status": "SUCCESS",
  "statusCode": 0,
}
`;

exports[`Encapsulation Test Common Packet Format Helper Functions Build Helper Function Generates Correct Output 1`] = `
Object {
  "data": Array [
    2,
    0,
    0,
    0,
    0,
    0,
    178,
    0,
    11,
    0,
    104,
    101,
    108,
    108,
    111,
    32,
    119,
    111,
    114,
    108,
    100,
  ],
  "type": "Buffer",
}
`;

exports[`Encapsulation Test Common Packet Format Helper Functions Build Helper Function Generates Correct Output 2`] = `
Object {
  "data": Array [
    3,
    0,
    0,
    0,
    0,
    0,
    178,
    0,
    11,
    0,
    104,
    101,
    108,
    108,
    111,
    32,
    119,
    111,
    114,
    108,
    100,
    161,
    0,
    14,
    0,
    84,
    104,
    105,
    115,
    32,
    105,
    115,
    32,
    97,
    32,
    116,
    101,
    115,
    116,
  ],
  "type": "Buffer",
}
`;

exports[`Encapsulation Test Common Packet Format Helper Functions Parse Helper Function Generates Correct Output 1`] = `
Array [
  Object {
    "TypeID": 0,
    "data": Object {
      "data": Array [],
      "type": "Buffer",
    },
  },
  Object {
    "TypeID": 178,
    "data": Object {
      "data": Array [
        104,
        101,
        108,
        108,
        111,
        32,
        119,
        111,
        114,
        108,
        100,
      ],
      "type": "Buffer",
    },
  },
]
`;

exports[`Encapsulation Test Common Packet Format Helper Functions Parse Helper Function Generates Correct Output 2`] = `
Array [
  Object {
    "TypeID": 0,
    "data": Object {
      "data": Array [],
      "type": "Buffer",
    },
  },
  Object {
    "TypeID": 178,
    "data": Object {
      "data": Array [
        104,
        101,
        108,
        108,
        111,
        32,
        119,
        111,
        114,
        108,
        100,
      ],
      "type": "Buffer",
    },
  },
  Object {
    "TypeID": 161,
    "data": Object {
      "data": Array [
        84,
        104,
        105,
        115,
        32,
        105,
        115,
        32,
        97,
        32,
        116,
        101,
        115,
        116,
      ],
      "type": "Buffer",
    },
  },
]
`;

exports[`Encapsulation Test Encapsulation Generator Functions Register Session Returns Correct Encapsulation String 1`] = `
Object {
  "data": Array [
    101,
    0,
    4,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`Encapsulation Test Encapsulation Generator Functions SendRRData Returns Correct Encapsulation String 1`] = `
Object {
  "data": Array [
    111,
    0,
    27,
    0,
    145,
    129,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    10,
    0,
    2,
    0,
    0,
    0,
    0,
    0,
    178,
    0,
    11,
    0,
    104,
    101,
    108,
    108,
    111,
    32,
    119,
    111,
    114,
    108,
    100,
  ],
  "type": "Buffer",
}
`;

exports[`Encapsulation Test Encapsulation Generator Functions SendUnitData Returns Correct Encapsulation String 1`] = `
Object {
  "data": Array [
    112,
    0,
    33,
    0,
    145,
    129,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    2,
    0,
    161,
    0,
    4,
    0,
    145,
    125,
    0,
    0,
    177,
    0,
    13,
    0,
    200,
    1,
    104,
    101,
    108,
    108,
    111,
    32,
    119,
    111,
    114,
    108,
    100,
  ],
  "type": "Buffer",
}
`;

exports[`Encapsulation Test Encapsulation Generator Functions Unregister Session Returns Correct Encapsulation String 1`] = `
Object {
  "data": Array [
    102,
    0,
    0,
    0,
    145,
    129,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;
